transport/socks5/stream_dialer.go

```diff
@@ -55,78 +86,144 @@ func (c *streamDialer) DialStream(ctx context.Context, remoteAddr string) (trans
 	}()
 
 	// For protocol details, see https://datatracker.ietf.org/doc/html/rfc1928#section-3
+	// Creating a single buffer for method selection, authentication, and connection request
+	// Buffer large enough for method, auth, and connect requests with a domain name address.
+
+	var buffer []byte
+
+	if c.credentials == nil {
+		// Method selection part: VER = 5, NMETHODS = 1, METHODS = 0 (no auth)
+		// +----+----------+----------+
+		// |VER | NMETHODS | METHODS  |
+		// +----+----------+----------+
+		// | 1  |    1     | 1 to 255 |
+		// +----+----------+----------+
+		header := [3 + 3 + 256 + 2]byte{}
+		buffer = append(header[:0], 5, 1, 0)
+	} else {
+		// https://datatracker.ietf.org/doc/html/rfc1929
+		// Method selection part: VER = 5, NMETHODS = 1, METHODS = 2 (username/password)
+		header := [3 + 3 + 255 + 255 + 3 + 256 + 2]byte{}
+		buffer = append(header[:0], 5, 1, 2)
+
+		// Authentication part: VER = 1, ULEN, UNAME, PLEN, PASSWD
+		// +----+------+----------+------+----------+
+		// |VER | ULEN |  UNAME   | PLEN |  PASSWD  |
+		// +----+------+----------+------+----------+
+		// | 1  |  1   | 1 to 255 |  1   | 1 to 255 |
+		// +----+------+----------+------+----------+
+		buffer = append(buffer, 1) // Auth version
+		buffer = append(buffer, byte(len(c.credentials.username)))
+		buffer = append(buffer, c.credentials.username...)
+		buffer = append(buffer, byte(len(c.credentials.password)))
+		buffer = append(buffer, c.credentials.password...)
```

Comment:
Less verbose and more readable:
```suggestion
		buffer = append(buffer,
			1,
			byte(len(c.credentials.username)),
			c.credentials.username...,
			byte(len(c.credentials.password)),
			c.credentials.password...
		)
```